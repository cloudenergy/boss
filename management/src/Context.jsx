import React from 'react'
import { Subject } from 'rxjs-compat'

export const ActionContext = React.createContext(new Subject());
