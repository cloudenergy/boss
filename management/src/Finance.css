.audit-content {
  list-style-type: none;
}

thead.sticky-top {
  background-color: white;
}
