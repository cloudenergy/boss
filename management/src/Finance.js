
import React from 'react'
import {Observable} from 'rxjs-compat'
import * as r from 'ramda'
import './Finance.css'
import fuseOptFrom from './fuseOpt'
import Fuse from 'fuse.js'
import datef from 'dateformat'
import {AuditAction} from './Action'
import {AuditContext} from './Context'
import Confirm from './audit/Confirm'
import Edit from './audit/Edit'
import Table from './audit/Table'
import {modal,rest} from './utils-eff'

const {Context, Var} = AuditContext

const fuseOpt = fuseOptFrom(['fundChannel.project.name', 'account', 'subbranch', 'locate', 'fundChannel.status'])

const statusMap = {
  'PASSED': {color: 'text-success', text: '通过'},
  'DENY': {color: 'text-danger', text: '失败'},
  'PENDING': {color: 'text-warning', text:'待审核'}
}
const lensStatusMap = (status, val) => r.lensPath([status, val])

const payChannelTable = [{
  name: '项目名称',
  lens: r.view(r.lensPath(['fundChannel', 'project', 'name']))
},{
  name: '账户名',
  lens: r.view(r.lensPath(['fundChannel','name']))
},{
  name: '银行卡号/支付宝账号',
  lens: r.view(r.lensPath(['account']))
},{
  name: '支行名称',
  lens: r.view(r.lensProp('subbranch'))
},{
  name: '账户归属区域',
  lens: x =>
    r.pathOr('',['locate', 'province'], x) + r.pathOr('',['locate', 'city'], x) + r.pathOr('',['locate', 'district'],x)
},{
  name: '审核',
  lens: r.compose(status=><span className={r.view(lensStatusMap(status, 'color'))(statusMap)}>
    {r.view(lensStatusMap(status, 'text'))(statusMap)}
  </span>,
                  r.view(r.lensPath(['fundChannel', 'status'])))
},{
  name: '申请时间',
  lens: r.compose(x=>x && datef(Date.parse(x), 'yyyy年mm月dd日 HH:MM'),
                  r.view(r.lensPath(['fundChannel', 'createdAt'])))
},{
  name: '操作',
  lens: r.compose(id=>(
    <button onClick={e=> {
      e.stopPropagation();
      modal('finance-edit', 'show')
      Var.next(AuditAction.Edit(id))
    }} className="btn btn-link">编辑</button>),
                  r.prop('fundChannelId'))
}]

const contextValue = {
  actions: Var,
  table:payChannelTable,
  idLens: r.view(r.lensPath(['fundChannel', 'id'])),
  statusLens: r.compose(r.equals("PENDING"), r.view(r.lensPath(['fundChannel', 'status']))),
  modalId: "auditing",
}

export default class Finance extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      payChannel: [],
      auditId: '',
      auditEnable: true,
      query: "",
      fuse: new Fuse([], fuseOpt)
    }
  }
  componentDidMount() {
    this.subscription = Var.startWith(AuditAction.Load).flatMap(action => action.case({
      Load: () => rest('boss/payChannels')
        .map(({response})=>this.setState({
          payChannel: response.payChannel,
          fuse: new Fuse(response.payChannel, fuseOpt)
        })),
      Popup: (id,status) => Observable.of(this.setState({auditId: id, auditEnable: status})),
      Edit: (id) => Observable.of(this.setState({auditId: id})),
      Query: (str) => Observable.of(this.setState({query: str})),
      Approve: id => rest(`boss/fundChannels/${id}/status`, {
        method: 'PUT',
        body: {status: 'PASSED'},
      }).do(()=>Var.next(AuditAction.Load)),
      Deny: id => rest(`boss/fundChannels/${id}/status`, {
        method: 'PUT',
        body: {status: 'DENY'},
      }).do(()=>Var.next(AuditAction.Load)),
      Update: (form, originalData)=> {
        let payChannel = {
          account: form.get('account'),
          linkman: form.get('linkman')||'',
          subbrach: form.get('subbrach'),
          locate: {
            province: form.get('province'),
            city: form.get('city'),
            district: form.get('district'),
          }
        }
        let fundChannel = {
          name: form.get('name')
        }
        return Observable.zip(
          rest(`boss/fundChannels/${originalData.fundChannelId}`, {
            method: 'PUT',
            body: fundChannel,
          }),
          rest(`boss/payChannels/${originalData.id}`, {
            method: 'PUT',
            headers: {'Content-Type': 'application/json'},
            body: payChannel,
          })
        ).do(()=>Var.next(AuditAction.Load))
      },
    })).subscribe()
  }
  componentWillUnmount(){
    this.subscription.unsubscribe()
  }
  render() {
    let filtered = this.state.query? this.state.fuse.search(this.state.query): this.state.payChannel
    let selected = filtered.find(p=> r.path(['fundChannel', 'id'])(p)=== this.state.auditId )
    let editingTable = r.equals('alipay', r.path(['fundChannel', 'tag'], selected))?[
      {name: '账号类型',
       lens: r.always('支付宝'),
       disabled: true,
       key: 'tag',
      },
      {
        name: '账户名',
        lens: r.view(r.lensPath(['fundChannel','name'])),
        key: 'name',
      },{
        name: '支付宝账号',
        lens: r.view(r.lensPath(['account'])),
        key: 'account'
      },{
        name: '姓名',
        lens: r.view(r.lensProp('linkman')),
        key: 'linkman',
      }
    ]:[{
      name: '账号类型',
      lens: r.always('银行卡'),
      disabled: true,
      key: 'tag'
    },{
      name: '账户名',
      lens: r.view(r.lensPath(['fundChannel','name'])),
      key: 'name'
    },{
      name: '银行卡号',
      lens: r.view(r.lensPath(['account'])),
      key: 'account'
    },{
      name: '支行名称',
      lens: r.view(r.lensProp('subbranch')),
      key: 'subbranch'
    },{ name: '省',
        lens: r.path(['locate', 'province']),
        key: 'province',
    },
       { name: '城市',
         lens: r.path(['locate', 'city']),
         key: 'city',
       },{ name: '地区',
           lens: r.path(['locate', 'district']),
           key:'district'
    }]
    return (
      <Context.Provider value={contextValue}>
        <Confirm enable={this.state.auditEnable} data={selected} auditId={this.state.auditId} />
        <Edit modalId="finance-edit" data={selected} table={editingTable} auditId={this.state.auditId} />
        <div className="accordion" id="banking-audit">
          <div className="card">
            <div className="card-header bg-info">
              <h5 className="mb-0">
                银行卡审核
              </h5>
            </div>
            <div>
              <div className="card-body">
                <div className="form-group">
                  <input className="form-control col-2" type="search" placeholder="搜索" aria-label="Search" onChange={e=> Var.next(AuditAction.Query(e.target.value))} />
                </div>
                <Table data={filtered} />
              </div>
            </div>
          </div>
        </div>
      </Context.Provider>
    )
  }
}
