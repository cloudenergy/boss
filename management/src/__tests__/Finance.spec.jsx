import React from 'react';
import renderer from 'react-test-renderer';
import * as r from 'ramda'
import Enzyme, {mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {AuditContext,createActionContext} from '../Context'
import {Observable} from 'rxjs-compat'
jest.mock('rxjs/ajax')

import Finance from '../Finance';
const {Context} = AuditContext
Enzyme.configure({ adapter: new Adapter() })

const payChannel = {"payChannel":[{"locate":null,"id":2,"fundChannelId":11,"documentId":null,"documentType":1,"account":"123@gmail.com","subbranch":"","reservedmobile":"","linkman":"heheda","mobile":"","createdAt":"2018-05-28T05:59:01.000Z","updatedAt":"2018-05-28T05:59:01.000Z","deletedAt":null,"fundChannel":{"id":11,"flow":"pay","projectId":"999999999","category":"online","tag":"alipay","name":"heheda","status":"PASSED","createdAt":"2018-05-28T05:59:01.000Z","updatedAt":"2018-07-13T13:44:08.000Z","deletedAt":null,"project":{"id":"888888","logoUrl":null,"name":"杭州海兴电力科技股份有限公司","address":null,"description":null,"telephone":null,"createdAt":"2018-04-24T10:25:32.000Z","updatedAt":"2018-04-24T10:25:34.000Z","deletedAt":null}}},{"locate":null,"id":3,"fundChannelId":12,"documentId":null,"documentType":1,"account":"000000000","subbranch":"","reservedmobile":"","linkman":"dddddd","mobile":"","createdAt":"2018-06-07T06:09:39.000Z","updatedAt":"2018-06-07T06:09:39.000Z","deletedAt":null,"fundChannel":{"id":12,"flow":"pay","projectId":"77777777","category":"online","tag":"alipay","name":"xxxxxxx","status":"PASSED","createdAt":"2018-06-07T06:09:39.000Z","updatedAt":"2018-07-13T13:44:10.000Z","deletedAt":null,"project":{"id":"33333333","logoUrl":null,"name":"测试项目","address":null,"description":null,"telephone":null,"createdAt":"2018-03-05T06:00:23.000Z","updatedAt":"2018-03-05T06:00:27.000Z","deletedAt":null}}}]}

describe('<Finance/>', ()=>{
  let subject
  beforeEach(()=>{
    subject = (<Finance/>)
  })

  it('renders', () => {
    Observable.ajax.mockReturnValue(Observable.of({response: payChannel}))
    expect(renderer.create(subject).toJSON()).toMatchSnapshot()
  })
})
