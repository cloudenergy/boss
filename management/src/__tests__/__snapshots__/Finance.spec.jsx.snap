// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Finance/> renders 1`] = `
Array [
  <div
    aria-hidden="true"
    className="modal"
    id="auditing"
    role="dialog"
    tabIndex="-1"
>
    <div
        className="modal-dialog modal-lg"
        role="document"
    >
        <div
            className="modal-content"
        >
            <div
                className="modal-header"
            >
                <button
                    aria-label="Close"
                    className="close"
                    data-dismiss="modal"
                    type="button"
                >
                    <span
                        aria-hidden="true"
                    >
                        ×
                    </span>
                </button>
            </div>
            <div
                className="modal-body container"
            >
                <div
                    className="row"
                >
                    <div
                        className="col-3"
                    >
                        <img
                            alt="logo"
                            src="logo.png"
                            style={
                                Object {
                                    "width": "160px",
                                  }
                            }
                        />
                    </div>
                    <div
                        className="col-9 border-left"
                    >
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                项目名称
                                :
                            </dt>
                            <dd
                                className="col-8"
                            />
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                账户名
                                :
                            </dt>
                            <dd
                                className="col-8"
                            />
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                银行卡号/支付宝账号
                                :
                            </dt>
                            <dd
                                className="col-8"
                            />
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                支行名称
                                :
                            </dt>
                            <dd
                                className="col-8"
                            />
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                账户归属区域
                                :
                            </dt>
                            <dd
                                className="col-8"
                            >
                                
                            </dd>
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                审核
                                :
                            </dt>
                            <dd
                                className="col-8"
                            >
                                <span
                                    className={undefined}
                                />
                            </dd>
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                申请时间
                                :
                            </dt>
                            <dd
                                className="col-8"
                            />
                        </dl>
                        <dl
                            className="row"
                        >
                            <dt
                                className="col-4"
                            >
                                操作
                                :
                            </dt>
                            <dd
                                className="col-8"
                            >
                                <button
                                    className="btn btn-link"
                                    onClick={[Function]}
                                >
                                    编辑
                                </button>
                            </dd>
                        </dl>
                    </div>
                </div>
            </div>
            <div
                className="modal-footer"
            >
                <button
                    className="btn btn-success"
                    data-dismiss="modal"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                >
                    确认审核
                </button>
                <button
                    className="btn btn-danger"
                    data-dismiss="modal"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                >
                    审核失败
                </button>
            </div>
        </div>
    </div>
</div>,
  <div
    aria-hidden="true"
    className="modal"
    id="finance-edit"
    role="dialog"
    tabIndex="-1"
>
    <div
        className="modal-dialog modal-lg"
        role="document"
    >
        <div
            className="modal-content"
        >
            <div
                className="modal-header"
            >
                <button
                    aria-label="Close"
                    className="close"
                    data-dismiss="modal"
                    type="button"
                >
                    <span
                        aria-hidden="true"
                    >
                        ×
                    </span>
                </button>
            </div>
            <form
                onSubmit={[Function]}
            >
                <div
                    className="modal-body container"
                >
                    <div
                        className="row"
                    >
                        <div
                            className="col-3"
                        >
                            <img
                                alt="logo"
                                src="logo.png"
                                style={
                                    Object {
                                        "width": "160px",
                                      }
                                }
                            />
                        </div>
                        <div
                            className="col-9 border-left"
                        >
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    账号类型
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue="银行卡"
                                        disabled={true}
                                        name="tag"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    账户名
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="name"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    银行卡号
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="account"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    支行名称
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="subbranch"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    省
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="province"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    城市
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="city"
                                    />
                                </dd>
                            </dl>
                            <dl
                                className="row"
                            >
                                <dt
                                    className="col-4"
                                >
                                    地区
                                    :
                                </dt>
                                <dd
                                    className="col-8"
                                >
                                    <input
                                        defaultValue={undefined}
                                        disabled={undefined}
                                        name="district"
                                    />
                                </dd>
                            </dl>
                        </div>
                    </div>
                </div>
                <div
                    className="modal-footer"
                >
                    <input
                        className="btn btn-success"
                        type="submit"
                        value="保存"
                    />
                </div>
            </form>
        </div>
    </div>
</div>,
  <div
    className="accordion"
    id="banking-audit"
>
    <div
        className="card"
    >
        <div
            className="card-header bg-info"
        >
            <h5
                className="mb-0"
            >
                银行卡审核
            </h5>
        </div>
        <div>
            <div
                className="card-body"
            >
                <div
                    className="form-group"
                >
                    <input
                        aria-label="Search"
                        className="form-control col-2"
                        onChange={[Function]}
                        placeholder="搜索"
                        type="search"
                    />
                </div>
                <table
                    className="table table-hover"
                >
                    <thead
                        className="sticky-top"
                    >
                        <tr>
                            <th
                                scope="col"
                            >
                                项目名称
                            </th>
                            <th
                                scope="col"
                            >
                                账户名
                            </th>
                            <th
                                scope="col"
                            >
                                银行卡号/支付宝账号
                            </th>
                            <th
                                scope="col"
                            >
                                支行名称
                            </th>
                            <th
                                scope="col"
                            >
                                账户归属区域
                            </th>
                            <th
                                scope="col"
                            >
                                审核
                            </th>
                            <th
                                scope="col"
                            >
                                申请时间
                            </th>
                            <th
                                scope="col"
                            >
                                操作
                            </th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr
                            data-toggle="modal"
                            onClick={[Function]}
                        >
                            <td>
                                杭州海兴电力科技股份有限公司
                            </td>
                            <td>
                                heheda
                            </td>
                            <td>
                                123@gmail.com
                            </td>
                            <td>
                                
                            </td>
                            <td>
                                
                            </td>
                            <td>
                                <span
                                    className="text-success"
                                >
                                    通过
                                </span>
                            </td>
                            <td>
                                2018年05月28日 15:59
                            </td>
                            <td>
                                <button
                                    className="btn btn-link"
                                    onClick={[Function]}
                                >
                                    编辑
                                </button>
                            </td>
                        </tr>
                        <tr
                            data-toggle="modal"
                            onClick={[Function]}
                        >
                            <td>
                                测试项目
                            </td>
                            <td>
                                xxxxxxx
                            </td>
                            <td>
                                000000000
                            </td>
                            <td>
                                
                            </td>
                            <td>
                                
                            </td>
                            <td>
                                <span
                                    className="text-success"
                                >
                                    通过
                                </span>
                            </td>
                            <td>
                                2018年06月07日 16:09
                            </td>
                            <td>
                                <button
                                    className="btn btn-link"
                                    onClick={[Function]}
                                >
                                    编辑
                                </button>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </div>
    </div>
</div>,
]
`;
